bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/internal/OpenThermGatewaySocketConnector.java

```diff
@@ -0,0 +1,219 @@
+ *
+ * @author Arjen Korevaar - Initial contribution
+ * @author Arjan Mels - Improved robustness by re-sending commands, handling all message types (not only Boiler)
+ */
+@NonNullByDefault
+public class OpenThermGatewaySocketConnector implements OpenThermGatewayConnector {
+    private static final int COMMAND_RESPONSE_TIME_MILLISECONDS = 100;
+    private static final int COMMAND_TIMEOUT_MILLISECONDS = 5000;
+
+    private final Logger logger = LoggerFactory.getLogger(OpenThermGatewaySocketConnector.class);
+
+    private final OpenThermGatewayCallback callback;
+    private final String ipaddress;
+    private final int port;
+
+    private @Nullable PrintWriter writer;
+
+    private volatile boolean stopping;
+    private boolean connected;
+
+    private Map<String, Entry<Long, GatewayCommand>> pendingCommands = new ConcurrentHashMap<>();
+
+    public OpenThermGatewaySocketConnector(OpenThermGatewayCallback callback, String ipaddress, int port) {
+        this.callback = callback;
+        this.ipaddress = ipaddress;
+        this.port = port;
+    }
+
+    @Override
+    public void run() {
+        stopping = false;
+        connected = false;
+
+        logger.debug("Connecting OpenThermGatewaySocketConnector to {}:{}", this.ipaddress, this.port);
+
+        callback.connecting();
+
+        try (Socket socket = new Socket()) {
+            socket.connect(new InetSocketAddress(this.ipaddress, this.port), COMMAND_TIMEOUT_MILLISECONDS);
+            socket.setSoTimeout(COMMAND_TIMEOUT_MILLISECONDS);
+
+            connected = true;
+
+            callback.connected();
+
+            logger.debug("OpenThermGatewaySocketConnector connected");
+
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+                    PrintWriter wrt = new PrintWriter(socket.getOutputStream(), true)) {
+                // Make writer accessible on class level
+                writer = wrt;
+
+                sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintReport, "A"));
+                // Set the OTGW to report every message it receives and transmits
+                sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintSummary, "0"));
+
+                while (!stopping && !Thread.currentThread().isInterrupted()) {
+                    @Nullable
+                    String message = reader.readLine();
+
+                    if (message != null) {
+                        handleMessage(message);
+                    } else {
+                        logger.debug("Connection closed by OpenTherm Gateway");
+                        break;
+                    }
+                }
+
+                logger.debug("Stopping OpenThermGatewaySocketConnector");
+            } finally {
+                connected = false;
+
+                logger.debug("OpenThermGatewaySocketConnector disconnected");
+                callback.disconnected();
+            }
+        } catch (IOException ex) {
```

Comment:
I would expect an update of status from this `IOException` because now disconnect is called but the proper details of the problem are lacking. See also: `https://www.openhab.org/docs/developer/guidelines.html#f-logging`

> In general bindings should NOT log to error/warn if e.g. connections are dropped - this is considered to be an external problem and from a system perspective to be a normal and expected situation. The correct way to inform users about such events is to update the Thing status accordingl

Reply:
Added calling callback on IOException to update Thing status, and changed logging to debug.

Reply:
You removed the status update from the `disconnected` callback so the thing will not go `OFFLINE` when this exception is caught. 

Also, there is no context/message provided to the user, it would make sense to have a dedicated handler which passes the exception so that the message can be include when setting the thing status `OFFLINE` with a `COMMUNICATION_ERROR`. These statuses are described in the [documentation](https://www.openhab.org/docs/concepts/things.html#status-details).

I can also imagine that the host is completely wrong (empty?) or the configuration is wrong in other ways, bindings should set the status than to `CONFIGURATION_ERROR`. I suppose in the old code the binding would just set it to OFFLINE without context or warning for the user which is not that user-friendly. Given it seems to be the first time that we mention this it can be a follow-up PR whenever have you time.

BTW: follow-up PR's of the original contributor do not require a second reviewer.

Reply:
So.. a rollback then on that last commit.. since now the logic is pretty messed up setting the Thing status.

Reply:
Reverted last commit. I should probably redesign the Thing status code, though it's unclear as to what this status should represent. If it's about being able to communicate with the OpenTherm gateway device, then having a connection or not should be enough to indicate an online or offline status. Configuration errors are not that relevant, because the host is a required field and whether the host is valid or not is not something the binding would hardly be able to verify. It is either able to connect or not. The hostname itself can be anything, a simple hostname, an ip address, an fqdn.

Reply:
@ArjenKorevaar shall we keep the status handling for a separate PR?